flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/sink/SocketClientSink.java

```diff
@@ -88,6 +136,7 @@ public void invoke(IN value) {
 	 */
 	private void closeConnection(){
 		try {
+			isRunning = false;
```

Comment:
You should wake-up the waiting thread here. Otherwise waiting on a lock does not make any sense.